EIPS/eip-131.md

```diff
@@ -0,0 +1,85 @@
+---
+eip: 131
+title: Add precompiled BLAKE2b contract
+author: Tjaden Hess (@tjade273), Jay Graber, (@carterpy), James Hancock (@madeoftin)
+discussions-to: https://ethereum-magicians.org/t/blake2b-f-precompile/3157
+status: Draft
+type: Standards Track
+category: Core
+created: 2016-06-30
+---
+
+## Simple Summary
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the EIP.-->
+Special Projects wishes greater Ethereum interopability with Zcash, IPFS, and Handshake. Coincidentally, all of these projects use the Blake2 hash function. So, this is a request to add the Blake2b F precompile 11 to the Istanbul hardfork. The first special project here will probably be creating a wrapped ZEC (WZEC) within Ethereum as well as wrapped Ether within Zcash. After that, some yet-to-be-determined bridge architecture will allow Ethereum to benefit from Zcash’s shielded transactions.
+
+## Abstract
+
+This EIP introduces a new precompiled contract which implements the BLAKE2b cryptographic hashing algorithm.
+
+## Motivation
+
+BLAKE2b is a useful addition to Ethereum's cryptographic primitives. BLAKE2b is faster and more efficient than the SHA hash functions, making it a good complement or alternative to SHA-3. The BLAKE2b algorithm is highly optimized for 64-bit CPUs, and is faster than MD5 on modern processors. Its inclusion will also improve interoperability between the Zcash blockchain and the EVM, and assist in future integration with IPFS, as they are working on switching their default hash function to BLAKE2b.
+
+One BLAKE2 digest in Solidity, (see https://github.com/tjade273/eth-blake2/blob/optimise_everything/contracts/blake2.sol) currently requires `~480,000 + ~90*INSIZE` gas, and a single verification of an [Equihash](https://www.cryptolux.org/images/b/b9/Equihash.pdf) PoW verification requires 2<sup>5</sup> to 2<sup>7</sup> iterations of the hash function, making use cases such as verification of Zcash block headers prohibitively expensive. However, an efficient implementation of BLAKE2b could allow verification of the Equihash PoW used in Zcash, making a BTC Relay - style SPV client possible on Ethereum.
+
+Adding a precompile for the BLAKE2b hashing function will make the operation significantly faster and cheaper, improving usability.
+
+## Parameters
+
+* `METROPOLIS_FORK_BLKNUM`: TBD
+* `GBLAKEBASE`: 30
+* `GBLAKEWORD`: 6
+
+## Specification
+
+Adds a precompiled contract for the blake2b hash function.
+
+Address of `0x9`
+
+Function accepts a variable length input interpreted as:
+
+    [OUTSIZE, D_1, D_2, ..., D_INSIZE]
+
+where `INSIZE` is the length in bytes of the input. Throws if `OUTSIZE` is greater than 64. Returns the `OUTSIZE`-byte BLAKE2b digest, as defined in [RFC 7693](https://tools.ietf.org/html/rfc7693).
+
+### Gas Costs
+
+Gas costs would be equal to `GBLAKEBASE + GBLAKEWORD * floor(INSIZE / 32)`
```

Comment:
Should gas depend on the value of `OUTSIZE`?

Reply:
We are digging further into this, and we are not sure yet. We found a golang library for blake2b and are going to do some benchmarking on it. I'll update the specification to say TBD based on benchmarking. https://github.com/golang/crypto/tree/master/blake2b